cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs

```diff
@@ -405,155 +406,149 @@ runQueryKesPeriodInfo (AnyConsensusModeParams cModeParams) network nodeOpCertFil
+      liftIO $ renderOpCertNodeAndOnDiskCounterInformation nodeOpCertFile counterInformation
+
+      let qKesInfoOutput = createQueryKesPeriodInfoOutput opCertIntervalInformation counterInformation eInfo gParams
+          kesPeriodInfoJSON = encodePretty qKesInfoOutput
+      case mOutFile of
+        Just (OutputFile oFp) -> do
+          liftIO $ LBS.putStrLn kesPeriodInfoJSON
+          handleIOExceptT (ShelleyQueryCmdWriteFileError . FileIOError oFp)
+            $ LBS.writeFile oFp kesPeriodInfoJSON
+        Nothing -> liftIO $ LBS.putStrLn kesPeriodInfoJSON
+
+
 
     mode -> left . ShelleyQueryCmdUnsupportedMode $ AnyConsensusMode mode
  where
-   -- Returns the number of slots left until KES key rotation and the KES period.
-   -- We check the calculated current KES period based on the current slot and
-   -- protocolParamSlotsPerKESPeriod matches what is specified in the 'OperationalCertificate'.
-   opCertKesPeriodCheck
-     :: OperationalCertificate
+   currentKesPeriod :: ChainTip -> GenesisParameters -> CurrentKesPeriod
+   currentKesPeriod ChainTipAtGenesis _ = CurrentKesPeriod 0
+   currentKesPeriod (ChainTip currSlot _ _) gParams =
+     let slotsPerKesPeriod = fromIntegral $ protocolParamSlotsPerKESPeriod gParams
+     in CurrentKesPeriod $ unSlotNo currSlot `div` slotsPerKesPeriod
+
+   opCertStartingKesPeriod :: OperationalCertificate -> OpCertStartingKesPeriod
+   opCertStartingKesPeriod = OpCertStartingKesPeriod . fromIntegral . getKesPeriod
+
+   opCertEndKesPeriod :: GenesisParameters -> OperationalCertificate -> OpCertEndingKesPeriod
+   opCertEndKesPeriod gParams oCert =
+     let OpCertStartingKesPeriod start = opCertStartingKesPeriod oCert
+         maxKesEvo = fromIntegral $ protocolParamMaxKESEvolutions gParams
+     in OpCertEndingKesPeriod $ start + maxKesEvo
+
+   -- See OCERT rule
+   opCertIntervalInfo
+     :: GenesisParameters
      -> ChainTip
+     -> CurrentKesPeriod
+     -> OpCertStartingKesPeriod
+     -> OpCertEndingKesPeriod
+     -> OpCertIntervalInformation
+   opCertIntervalInfo gParams currSlot' c s e@(OpCertEndingKesPeriod oCertEnd) =
+       let cSlot = case currSlot' of
+                       (ChainTip cSlotN _ _) -> unSlotNo cSlotN
+                       ChainTipAtGenesis -> 0
+           slotsTillExp = SlotsTillKesKeyExpiry . SlotNo $ (oCertEnd * fromIntegral (protocolParamSlotsPerKESPeriod gParams)) - cSlot
+       in O.createOpCertIntervalInfo c s e (Just slotsTillExp)
+
+   opCertNodeAndOnDiskCounters
+     :: OpCertOnDiskCounter
+     -> Maybe OpCertNodeStateCounter
+     -> OpCertNodeAndOnDiskCounterInformation
+   opCertNodeAndOnDiskCounters o@(OpCertOnDiskCounter odc) (Just n@(OpCertNodeStateCounter nsc))
+     | odc < nsc = OpCertOnDiskCounterBehindNodeState o n
+     | otherwise = OpCertOnDiskCounterMoreThanOrEqualToNodeState o n
+   opCertNodeAndOnDiskCounters o Nothing = OpCertNoBlocksMintedYet o
```

Comment:
What does the node state counter and on disk counter have to do with whether or blocks are minted yet?

Reply:
Have a look at `currentIssueNo`: https://github.com/input-output-hk/cardano-ledger/blob/ac51494e151af0ad99b937a787458ce71db0aaea/libs/cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/OCert.hs#L67
We currently only check the counter map for the existence of the corresponding stake pool's operational certificate counter. If it's not in the counter map it means the stake pool hasn't minted a block yet.